3rd party/AutoResourceStr/variantWrapper.pas

```diff
@@ -0,0 +1,249 @@
+
+procedure TAbstractWrapperData.Release;
+begin
+  Free;
+//  Destroy;  //can be overriden to put this object into object pool where it can be re-used
+end;
+
+class function TAbstractWrapperData.getInstance: TAbstractWrapperData;
+begin
+  Result:=Create; //can be overriden to take new instance from object pool if available
+  //it may be very efficient indeed.
+end;
+
+procedure TAbstractWrapperData.DoSubtract(const Right: TAbstractWrapperData);
+var WrapperClass: TWrapperDataClass;
+    Clone: TAbstractWrapperData;
+begin
+  //here we expect that A-B = A+(-B)
+  //in most cases it's correct.
+  //what's interesting, we're not expecting that - (-B) = B.
+  //previous implementation didn't copy object but just negated it 2 times,
+  //but not sure it never affects Right when working with floating point for ex.
+  WrapperClass:=TWrapperDataClass(Right.ClassType);
+  Clone:=WrapperClass.GetInstance;
+  try
+    Clone.FastAssign(Right);
+    Clone.DoNegate;
+    DoAdd(Clone);
+  finally
+    Clone.Release;
+  end;
+end;
+
+procedure TAbstractWrapperData.FastAssign(source: TAbstractWrapperData);
+begin
+  Assign(source); //not so fast here, but can be overriden
+end;
+
+(*
+    TAbstractWrapperVariantType
+                                  *)
+function TAbstractWrapperVariantType.GetInstance(const V: TVarData): TObject;
+begin
+  Result:=TWrapperVarData(V).Data;
+end;
+
+procedure TAbstractWrapperVariantType.Clear(var V: TVarData);
+begin
+  V.VType:=varEmpty;
+  TWrapperVarData(V).Data.Release;
+end;
+
+procedure TAbstractWrapperVariantType.Copy(var Dest: TVarData; const Source: TVarData; const Indirect: Boolean);
+var WrapperClass: TWrapperDataClass;  //�������� � Create ����������� ������
+begin
+//it seems that custom variant is never copied indirectly,
+//(at least as it's implemented in variants.pas for all Delphi versions up to 10 Seattle)
+//that's why commented this part for better perfomance
+
+//  if Indirect and VarDataIsByRef(Source) then
+//    VarDataCopyNoInd(Dest, Source)
+//  else
+// ��� custom variant'� ������� �� ����� �������� �� ������
+    with TWrapperVarData(Dest) do
+    begin
+      VType := VarType;
+      WrapperClass:=TWrapperDataClass(TWrapperVarData(Source).Data.ClassType);
+      Data:=WrapperClass.GetInstance;  //����� ��� �������� ����������� VariantType ������� WrapperClass!
+      Data.FastAssign(TWrapperVarData(Source).Data);
+    end;
+end;
+
+procedure TAbstractWrapperVariantType.UnaryOp(var Right: TVarData; const Operator: Integer);
+begin
+//������� ����� �, ��������, ���������� not.
+  if Right.vtype=VarType then
+    if Operator=opNegate then
+      TWrapperVarData(Right).Data.DoNegate
+    else
+      RaiseInvalidOp
+  else
+    RaiseInvalidOp;
+end;
+
+procedure TAbstractWrapperVariantType.BinaryOp(var Left: TVarData; const Right: TVarData; const Operator: TVarOp);
+//var CastedRight: TVarData;
+begin
+//�������, �������, ��������, ��������, ������� �� ������� � ������� �������� (������, ���. � ��)
```

Comment:
That looks a bit odd. Can you remove all the outcommented code and non-english comments? It's just adding noise.